### PR TITLE
daemon: allow enabling enforce mode

### PR DESCRIPTION
**Pull request by stolowski**
Allow enabling enforcing mode for a validation set with `snap validate --enforce ...`.

The existing spread test is only updated not to fail. Actual enforcing will be spread-tested separately once all the bits land.

### CODE REVIEW

File: daemon/api_validate.go [hunk 1]
```diff
@@ -346,19 +353,27 @@ func validationSetForAssert(st *state.State, accountID, name string, sequence in
 	return sets, nil
 }
 
-func validationSetAssertFromDb(st *state.State, accountID, name string, sequence int) (*asserts.ValidationSet, error) {
+func validationSetAssertFromDb(st *state.State, accountID, name string, sequence int) (vset *asserts.ValidationSet, err error) {
```
Comment (mardy) [inline daemon/api_validate.go:356]: At a first sight, the changes to this function seem to be unrelated from the rest. If it's so, I'd rather have them in a separate commit explaining what changed (or maybe you can expand the text in the current commit, if that's too bothersome).

Comment (stolowski) [reply daemon/api_validate.go]: Yes, you right, thanks for spotting. This was unintended change, a result of having a few WIP branches where I take stuff from for this feature. Removed.

**Review by mardy — APPROVED** [review]
LGTM!

File: daemon/api_validate.go [hunk 2]
```diff
@@ -414,3 +422,27 @@ func getSingleSeqFormingAssertion(st *state.State, accountID, name string, seque
...
+func enforceValidationSet(st *state.State, accountID, name string, sequence, userID int) Response {
+	snaps, err := installedSnaps(st)
+	if err != nil {
+		return InternalError(err.Error())
+	}
+	vs, err := validationSetAssertionForEnforce(st, accountID, name, sequence, userID, snaps)
+	if err != nil {
+		// XXX: provide more specific error kinds? This would probably require
+		// assertstate.ValidationSetAssertionForEnforce tuning too.
+		return BadRequest("cannot enforce validation set: %v", err)
```
Comment (pedronis) [inline daemon/api_validate.go:435]: in the spread test we will want to check for an error case too, and make sure the final error the user sees is not too repetitive

File: daemon/api_validate_test.go [hunk 3]
```diff
@@ -926,3 +919,189 @@ func (s *apiValidationSetsSuite) TestApplyValidationSetUnsupportedAction(c *chec
...
+	c.Assert(err, check.IsNil)
+	c.Check(tr, check.DeepEquals, assertstate.ValidationSetTracking{
+		Mode:      assertstate.Enforce,
+		AccountID: s.dev1acct.AccountID(),
+		Name:      "bar",
+		Current:   3,
+	})
+}
+
+func (s *apiValidationSetsSuite) TestApplyValidationSetEnforceModeError(c *check.C) {
```
Comment (pedronis) [inline daemon/api_validate_test.go:1073]: do we need an explicit test where we are in monitor mode but switching to enforce fails so we stay in monitor mode?

**Review by pedronis — APPROVED** [review]
lgtm, couple small comments

File: daemon/api_validate_test.go (hunk 3 above)
Comment (stolowski) [reply daemon/api_validate_test.go:1073]: Good idea, will add one.

File: daemon/api_validate.go (hunk 2 above)
Comment (stolowski) [reply daemon/api_validate.go:435]: It's not ideal:
```
# snap validate --enforce xSfWKGdLoQBoQx88vIM1MpbFNMq53t1f/bar=1
error: cannot apply validation set: cannot enforce validation set: validation sets assertions are not met:
- missing required snaps:
  - hello-world (required by sets xSfWKGdLoQBoQx88vIM1MpbFNMq53t1f/bar)
```

but I think making it better needs a few changes in other places and is worth a PR on its own?

**Comment by mvo5** [comment]
Which PRs are left before we can remove the "blocked" label here?

**Comment by stolowski** [comment]
> Which PRs are left before we can remove the "blocked" label here?

This isn't blocked anymore, the required PRs about enforcing on refresh/install landed.
